Games/types/Mafia/items/Gun.js

```diff
@@ -10,53 +10,75 @@ module.exports = class Gun extends Item {
-                        else
-                            this.game.queueAlert(`:sy0a: Someone fires a gun at ${this.target.name}!`);
-
-                        // kill
-                        var toKill = this.dominates();
-                        if (mafiaImmune && this.target.role.alignment == "Mafia")
-                            toKill = false
-
-                        if (toKill) {
-                            this.target.kill("gun", this.actor, true);
-                        }
-
-                        this.item.drop();
-                        this.game.broadcast("gunshot");
+        this.baseMeetingName = "Shoot Gun";
+        this.currentMeetingIndex = 0;
+
+        this.gunMeeting = {
+            actionName: "Shoot",
+            states: ["Day"],
+            flags: ["voting", "instant", "noVeg"],
+            action: {
+                labels: ["kill", "gun"],
+                item: this,
+                run: function () {
+                    var shooterMask = this.actor.role.data.shooterMask;
+                    var reveal = shooterMask ? true : this.item.reveal;
+                    if (reveal == null) {
+                        reveal = Random.randArrayVal([true, false]);
+                    }
+                    if (shooterMask == null) {
+                        shooterMask = this.actor.name;
+                    }
+                    
+                    var mafiaImmune = this.item.mafiaImmune;
+                    var cursed = this.item.cursed;
+
+                    if (cursed) {
+                        this.target = this.actor;
+                    }
+
+                    if (reveal && cursed)
+                        this.game.queueAlert(`:sy0b: ${shooterMask} pulls a gun, it backfires!`);
+                    else if (reveal && !cursed)
+                        this.game.queueAlert(`:sy0a: ${shooterMask} pulls a gun and shoots at ${this.target.name}!`);
+                    else
+                        this.game.queueAlert(`:sy0a: Someone fires a gun at ${this.target.name}!`);
+
+                    // kill
+                    var toKill = this.dominates();
+                    if (mafiaImmune && this.target.role.alignment == "Mafia")
+                        toKill = false
+                    
+                    let toSnatch = false;
+                    // snatch gun
+                    if (!cursed && 
+                        (this.target.role.name == "Gunslinger" || this.target.role.name == "Sharpshooter") &&
+                        Random.randFloatRange(0, 100) <= 80) {
+                        toSnatch = true;
+                        toKill = false;
+                    }
+
+                    if (toKill) {
+                        this.target.kill("gun", this.actor, true);
+                    }
+
+                    this.item.drop();
+                    this.game.broadcast("gunshot");
+
+                    if (toSnatch) {

```

Comment:
should there be a system message here? i like it when it's ambiguous, it can mean vest/ granny/ gramps etc

Reply:
Gun immunity should be a feature of the roles not the gun